### PR TITLE
fix(server): key JIT account linking on upstream (issuer, subject)

### PR DESCRIPTION
**Pull request by jplock**
## Context

Prompted by ["Stop using email as a primary key"](https://workos.com/blog/stop-using-email-as-a-primary-key). Vouch linked upstream IdP logins to local accounts **by email alone** — the OIDC `sub` claim was never parsed (`IdTokenClaims` had no `sub` field), and an email match returned the pre-existing account verbatim. `complete_enrollment_after_identity` then minted a full-scope session and let the caller register a **new** security key on that account. So an email reassigned/recycled at a trusted IdP (or asserted by a second configured IdP) yielded takeover of the existing account — the article's "NoAuth" failure mode.

## What this changes

Bind accounts to the durable upstream identity — the validated OIDC `(iss, sub)` pair, or for SAML the IdP entity ID plus NameID — and treat email as mutable profile data.

- **`UserDoc.idp_identities`**: a per-issuer binding list (`Vec<IdpIdentity{issuer, subject}>`, one per issuer), indexed as a blind `idp_identity` point lookup. Additive with `#[serde(default)]` — legacy documents deserialize to an empty list, no schema migration.
- **Matching in `enroll_user_with_org`** (now `Result<EnrolledUser, EnrollUserError>`):
  1. Binding match on `(issuer, subject)` wins, regardless of asserted email.
  2. Email match with **no binding for this issuer** → lazily bind on first login (`newly_bound`). No batch backfill: legacy and SCIM-provisioned accounts bind on their first IdP login.
  3. Email match whose binding for this issuer names a **different subject** → refused with `IdentityConflict` (the reassignment/takeover shape).
  4. No match → new account created carrying the binding.
- **OIDC**: `sub` is now required (fail-closed if absent) and the binding pins the concrete token issuer (`claims.iss`), correct for the Entra `/organizations` `{tenantid}` template.
- **SAML**: parser captures NameID + Format; transient-format NameIDs skip binding (email-only fallback, logged) so those deployments aren't bricked.
- **Email is canonical**: a drifted upstream email is logged but never written back; downstream uses in `complete_enrollment_after_identity` switched to the stored `user.email`.
- **Refusal** renders an i18n'd "Account Linking Blocked" page and emits an audit event.
- **Audit**: new `identity_bound` / `identity_bind_refused` kinds (AuthEvents retention, OCSF-mapped). Payloads carry `idp_issuer` only — never the raw subject (a SAML NameID is often an email).

## Concurrency

Lazy bind uses `compare_and_update` + `OccConflict` under the existing `with_dsql_retry!`, the same pattern as the org-admin claim in this function — never a blind update.

## Tests

- DB-level: binding wins over email (+ no email rewrite on drift); same-issuer/different-subject refused; lazy bind on a legacy account (incl. a second issuer); new-user creation binds; SCIM user binds on first IdP login.
- Handler-level: refusal renders an error, sets no session cookie, and audits `identity_bind_refused` with `idp_issuer` and no raw email; lazy bind emits `identity_bound`.
- Unit: OIDC `sub` required/returned; SAML NameID + Format extraction; `UserDoc` serde default and index entries.
- `make fmt`, `cargo clippy --all-targets --all-features -D warnings`, full `vouch-server` suite (2722), `vouch-tests`, and the `audit_event_docs` / `no_raw_email_in_audit_data` / `arch_boundaries` gates all pass. `prek run` clean.

## Docs

`docs/src/idp/overview.md` (account-linking/identity-binding model, recovery via admin remove + re-enroll), `docs/src/idp/saml.md` (NameID stability guidance), `docs/src/admin/audit.md` (two new event rows).

## Known limitations (documented)

- No unbind surface: a legitimately re-issued subject locks the user out until an admin removes them and they re-enroll.
- Transient SAML NameID deployments stay email-only (warned + documented).
- Cloud-federation tokens still use `sub = email` (separate issue; out of scope here).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
_Generated by [Claude Code](https://claude.ai/code/session_016ryqP5bDWAhSPsND2gYRs1)_

---

> [!NOTE]
> **High Risk**
> Changes core authentication account resolution and session issuance; mistakes could lock out users or still allow wrong-account linking, though behavior is heavily tested and fail-closed on conflicts.
> 
> **Overview**
> Closes the **email-only JIT linking** takeover path by binding Vouch accounts to upstream **(issuer, subject)** instead of treating email as the durable identity.
> 
> **Data & matching:** `UserDoc` gains `idp_identities` (one binding per issuer) with a blind `idp_identity` index; legacy rows deserialize with an empty list. `enroll_user_with_org` now takes optional `UpstreamLogin`, resolves **binding → lazy bind on email → refuse on issuer conflict → create**, and returns `EnrollUserError::IdentityConflict` when the same issuer cannot reassert the bound subject (including non-durable SAML logins after a persistent bind).
> 
> **Protocols:** OIDC requires `sub` and pins binding to validated `claims.iss` (Entra tenant template). SAML captures NameID + Format; only **persistent** NameIDs become `durable_subject`.
> 
> **Enrollment UX & audit:** Conflicts show **Account Linking Blocked** (i18n); successful lazy binds emit `identity_bound`, refusals emit `identity_bind_refused` (payload: `idp_issuer` only). Sessions and downstream flows use stored `user.email`; upstream email drift is logged, not written back.
> 
> Docs cover the linking model, SAML NameID guidance, and new audit/OCSF event types. Broad DB/handler tests cover binding, refusal, SCIM first login, and SAML format edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 65c974e486ae6a50d65f388d481564d2350eaf36. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: crates/vouch-server/src/handlers/saml.rs [hunk 1]
```diff
@@ -169,9 +169,30 @@ pub(crate) async fn acs(
...
+                idp = %saml_provider.id,
+                "SAML NameID format is transient; skipping identity binding \
+                 (account matching falls back to email only)"
+            );
+            None
+        }
+        (Some(name_id), _) => Some(crate::db::IdpIdentity {
+            issuer: saml_provider.idp_metadata.entity_id.clone(),
+            subject: name_id.clone(),
+        }),
```
[the remark this answers is not in the document]
Comment (jplock) [reply crates/vouch-server/src/handlers/saml.rs]: Good catch — fixed in 5cc40ef. The transient-only skip was too narrow. Binding is now an allowlist: only `persistent` and `emailAddress` NameID formats are bound; transient, `unspecified`, unknown URNs, and an absent `Format` all fall back to email-only matching (logged), so a rotating NameID can no longer lazily bind on the first login and then lock the user out on the second. The decision is extracted into a `saml_upstream_identity` helper with unit tests for each format.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (jplock) [reply crates/vouch-server/src/handlers/saml.rs]: Fixed in 7d2f52a0 by inverting to a persistent-only allowlist: only `Format=persistent` NameIDs are eligible to create an `(issuer, subject)` binding. `emailAddress`, `unspecified`, `transient`, and a missing `Format` attribute all fall back to email-only matching (pre-binding behavior), so a rotating NameID under any of those formats can never create a binding or trigger the conflict refusal.

[the remark this answers is not in the document]
Comment (jplock) [reply crates/vouch-server/src/handlers/saml.rs]: Resolved in 5cc40efc and 7d2f52a0: binding is now allowlist-based — only a `Format=persistent` NameID creates an `(issuer, subject)` binding. `emailAddress`, `unspecified`, `transient`, unknown, and missing formats fall back to email-only account matching and never create a binding, so a rotating NameID can no longer cause a conflict lockout.

File: crates/vouch-server/src/handlers/saml.rs [hunk 2]
```diff
@@ -184,6 +198,40 @@ pub(crate) async fn acs(
...
+    name_id_format: Option<&str>,
+) -> Option<crate::db::IdpIdentity> {
+    let name_id = name_id?;
+    if name_id_format != Some(NAMEID_PERSISTENT) {
+        return None;
+    }
+    Some(crate::db::IdpIdentity {
+        issuer: entity_id.to_string(),
+        subject: name_id.to_string(),
+    })
```
[the remark this answers is not in the document]
Comment (jplock) [reply crates/vouch-server/src/handlers/saml.rs]: Fixed in ee4770a8 (plus test coverage added in 65c974e4). Split the login's issuer from its durable subject (`UpstreamLogin { issuer, durable_subject }`) so the issuer is always known even when the NameID format isn't persistent. `enroll_user_with_org` now refuses an email-matched login when the account already holds a binding for that issuer and the login can't reassert it (absent or different subject), instead of skipping the check when `upstream` was `None`. Accounts with no existing binding for the issuer still match on email alone, so a rotating NameID still can't cause the original #837 lockout. Verified with cases for: bound account + non-durable login on the same issuer (refused + audited), never-bound account + non-durable login (still succeeds), and bound-for-a-different-issuer + non-durable login on the unrelated issuer (still succeeds).
